### PR TITLE
Fixed asserts with EvaluateAttributeAtSample

### PR DESCRIPTION
**Pull request by adam-yang**
An assert occurs in the following scenario:
```
void main( attr ) {
  if (...) {
    uv = EvaluateAttributeAtSample(attr, ...);
    Sample(uv);
  }
}
```

This happens because the `Sample()` call is inside of control flow, a convergence guard is created for `attr` outside of the if block. When lowering `EvaluateAttributeAtSample`, the args are expected to be `loadInput`, but convergence guard is seen instead.

The solution is to not create convergence guard for `Argument` values. Arguments get lowered to loadInput, which do not get sinked anyway.

### CODE REVIEW

File: lib/HLSL/HLOperationLower.cpp [hunk 1]
```diff
@@ -829,7 +829,11 @@ Constant *GetLoadInputsForEvaluate(Value *V, std::vector<CallInst*> &loadList) {
...
     while (!isa<UndefValue>(Vec)) {
-      InsertElementInst *insertInst = cast<InsertElementInst>(Vec);
+      InsertElementInst *insertInst = dyn_cast<InsertElementInst>(Vec);
+      if (!insertInst) {
+        Vec->getContext().emitError(kAttributeEvaluateError);
+        break;
+      }
       Vec = insertInst->getOperand(0);
       Value *Elt = insertInst->getOperand(1);
       if (IsValidLoadInput(Elt)) {
```
Comment (MrTrillian) [inline lib/HLSL/HLOperationLower.cpp]: Can there be a case where only some of the elements are loadinputs? Something like this?

float main(float foo : IN) : OUT {
  float2 bar = float2(foo, 42);
}

Comment (adam-yang) [reply lib/HLSL/HLOperationLower.cpp]: `EvaluateAttributeAtSample` must be given directly the attribute AFAIK.

**Comment by MrTrillian** [comment]
`batch/expressions/intrinsics/EvaluateAttributeAtSample` would be a better location. `misc` is the backlog of stuff to be triaged into the organized locations.

File: lib/HLSL/HLOperationLower.cpp (hunk 1 above)
Comment (python3kgae) [reply lib/HLSL/HLOperationLower.cpp]: Could it be Shuffle which generated from swizzle like foo.zx when foo is float4?

File: lib/HLSL/DxilConvergent.cpp [hunk 2]
```diff
@@ -156,11 +156,12 @@ bool DxilConvergentMark::PropagateConvergentImpl(Value *V, Function *F,
       // TODO: static indexing cbuffer is fine.
       return true;
     }
-  } else {
+  } else if (!isa<Argument>(V)) {
```
Comment (python3kgae) [inline lib/HLSL/DxilConvergent.cpp]: What will happen if argument also used directly as sample coordinate?

Texture2D tex0 : register(t0);
SamplerState samp0 : register(s0);

 float4 main(float2 u : TEXCOORD0, float v : TEXCOORD1) : SV_TARGET {
  float4 result = float4(0,0,0,0);
  int i = 0;
  [branch]
  if (u > 0.5) {
    float2 coord = float2(
      EvaluateAttributeAtSample(u.x, i),
      EvaluateAttributeAtSample(v, i)
    );
    result += tex0.Sample(samp0, u);
    result += tex0.Sample(samp0, coord);
  }
  return result;
}

File: lib/HLSL/HLOperationLower.cpp (hunk 1 above)
Comment (MrTrillian) [reply lib/HLSL/HLOperationLower.cpp]: @adam-yang It's incorrect but I was wondering how the code would behave in that case.

Comment (adam-yang) [reply lib/HLSL/HLOperationLower.cpp]: @python3kgae As it is now, we cannot do `EvaluateAttributeAtSample(foo.yx)`. Should this be allowed?

Comment (python3kgae) [reply lib/HLSL/HLOperationLower.cpp]: We can do it later.

File: lib/HLSL/DxilConvergent.cpp (hunk 2 above)
Comment (adam-yang) [reply lib/HLSL/DxilConvergent.cpp]: Turns out this doesn't work, loadInput gets sank. I think I'll do your suggested high level intrinsics solution.

**Pull request closed without merging**